src/core/dater.rs

```diff
@@ -0,0 +1,209 @@
+use lazy_static::lazy_static;
+
+use crate::core::matter::{tables as matter, Matter};
+use crate::error::{err, Error, Result};
+
+type Blake2b256 = blake2::Blake2b<blake2::digest::consts::U32>;
+
+pub trait Dater {
+    fn new_with_code_and_raw(code: &str, raw: &[u8]) -> Result<Matter>;
+    fn new_with_dts(dts: &str) -> Result<Matter>;
+    fn new_with_dtsb(dts: &[u8]) -> Result<Matter>;
+    fn new_with_qb64(qb64: &str) -> Result<Matter>;
+    fn new_with_qb64b(qb64b: &[u8]) -> Result<Matter>;
+    fn new_with_qb2(qb2: &[u8]) -> Result<Matter>;
+    fn dts(&self) -> Result<String>;
+    fn dtsb(&self) -> Result<Vec<u8>>;
+}
+
+fn validate_code(code: &str) -> Result<()> {
+    lazy_static! {
+        static ref CODES: Vec<&'static str> = vec![matter::Codex::DateTime.code(),];
+    }
+
+    if !CODES.contains(&code) {
+        return err!(Error::UnexpectedCode(code.to_string()));
+    }
+
+    Ok(())
+}
+
+fn iso_8601_to_b64(dts: &str) -> String {
+    dts.replace(':', "c").replace('.', "d").replace('+', "p")
+}
+
+fn b64_to_iso_8601(dts: &str) -> String {
+    dts.replace('c', ":").replace('d', ".").replace('p', "+")
+}
+
+fn now_as_iso8601() -> String {
+    let dt = chrono::offset::Utc::now();
+    dt.to_rfc3339_opts(chrono::SecondsFormat::Micros, false)
+}
+
+fn now_as_b64() -> String {
+    let dt = now_as_iso8601();
+    iso_8601_to_b64(&dt)
+}
+
+impl Dater for Matter {
+    fn new_with_code_and_raw(code: &str, raw: &[u8]) -> Result<Matter> {
+        if !code.is_empty() {
+            validate_code(code)?;
+        }
+        if raw.is_empty() {
+            let qb64 = format!("{}{}", matter::Codex::DateTime.code(), now_as_b64());
+            Matter::new_with_qb64(&qb64)
+        } else {
+            Matter::new_with_code_and_raw(code, raw)
+        }
```

Comment:
What happens here if code is empty and raw is not?

Reply:
Oh, you are relying on Matter to raise an error. I see.

Reply:
Yep. Tests check this case.